### ISSUE TITLE
quality issues on customized models

### ISSUE THREAD

**Issue by JunMa11** [wyhuai/DDNM#11]
Hi @wyhuai ,

Happy new year of the rabbit. 

I like your work very much. Recently, I tried it for medical image super-resolution and denoising but suffered from quality issues.

I trained a [DDPM](https://github.com/openai/guided-diffusion) on my dataset and verified that it can generate very good images. 

![ct_3](https://user-images.githubusercontent.com/19947331/214431024-b57b0730-5ca1-48ac-80ab-e192d0ddca09.png)

Then, I applied the DDNM for denoising and SR.

Original image:

128x128
![00000](https://user-images.githubusercontent.com/19947331/214430012-db07bae0-3101-4b4d-a05d-7fa2d8d3c652.png)

Reconstructed image

![00000](https://user-images.githubusercontent.com/19947331/214430115-47108a3c-1209-43c5-a6a9-57debd6f6ae0.png)

The results still have large noise and obvious block effects. I have tried different sigma_y but it didn't improve the quality.

I modified the code https://github.com/wyhuai/DDNM/blob/main/hq_demo/main.py as follows

Any comments are highly appreciated. BTW, the question is not urgent. Enjoy your holidays first:)

```python
"""
This is an DDNM implementation based on RePaint. 
This implementation support arbitrary in/out image size.
At present, we only implemented SR and noisy SR tasks.
"""

import os
import argparse
import torch as th
import torch.nn.functional as F
import time
import conf_mgt
from utils import yamlread
from guided_diffusion import dist_util
import torchvision.transforms as transforms
from PIL import Image

# Workaround
try:
    import ctypes
    libgcc_s = ctypes.CDLL('libgcc_s.so.1')
except:
    pass


from guided_diffusion.script_util import (
    NUM_CLASSES,
    model_and_diffusion_defaults,
    classifier_defaults,
    add_dict_to_argparser,
    create_model_and_diffusion,
    create_classifier,
    select_args,
    args_to_dict,
)  # noqa: E402

def read_model_and_diffusion(args, model_path):
    """Reads the latest model from the given directory."""

    model, diffusion = create_model_and_diffusion(
        **args_to_dict(args, model_and_diffusion_defaults().keys()), conf=conf,
    )
    model.load_state_dict(dist_util.load_state_dict(model_path, map_location="cuda"))

    # if args.use_fp16:
    #     model.convert_to_fp16()
    model.eval()
    return model, diffusion

# def main(conf: conf_mgt.Default_Conf, args):

# if __name__ == "__main__":
parser = argparse.ArgumentParser()
parser.add_argument('--config', type=str, required=False, default="/home/jma/Documents/I2I/DDNM/hq_demo/confs/ct256.yml")
parser.add_argument('--deg', type=str, required=False, default="sr_averagepooling")
parser.add_argument('--sigma_y', type=float, required=False, default=1.5)
parser.add_argument('-i', "--save_path", type=str, required=False, default="demo15") 

"""
SR scales should be divisible by 256, e.g., 2, 4, 8, 16 ...
"""    
parser.add_argument('--scale', type=int, required=False, default=4)


"""
resize y to the same shape with the desired result
"""
parser.add_argument('--resize_y', default=True, action='store_true')
parser.add_argument('--path_y', type=str, required=False, default="data/datasets/ct128.png")
"""
no classifier guidance
"""
parser.add_argument('--class', type=int, required=False, default=323)
args = vars(parser.parse_args())

conf_arg = conf_mgt.conf_base.Default_Conf()
conf_arg.update(yamlread(args.get('config')))

def create_argparser():
    defaults = dict(
        image_size=256,
        batch_size=1,
        num_channels=64,
        num_res_blocks=3,
        num_heads=1,
        diffusion_steps=1000,
        noise_schedule='linear',
        lr=1e-4,
        clip_denoised=False,
        num_samples=1, # 10000
        use_ddim=True,
        # timestep_respacing='ddim250',
        model_path="",
    )
    ori = model_and_diffusion_defaults()
    # defaults.update(model_and_diffusion_defaults())
    ori.update(defaults)
    parser = argparse.ArgumentParser()
    add_dict_to_argparser(parser, ori)
    return parser

model_args = create_argparser().parse_args()

conf = conf_arg

print("Start", conf['name'])
print(f"{conf=}")
print(f"{args=}")

os.chdir('/home/jma/Documents/I2I/DDNM/hq_demo')
ct_model_path = './data/pretrained/ct256.pt'
device = dist_util.dev(conf.get('device'))
model, diffusion = read_model_and_diffusion(model_args, ct_model_path)
model.to(device)
if conf.use_fp16:
    model.convert_to_fp16()
model.eval()

show_progress = conf.show_progress

cond_fn = None

def model_fn(x, t, y=None, gt=None, **kwargs):
    assert y is not None
    return model(x, t, y if conf.class_cond else None, gt=gt)

print("sampling...")
all_images = []

dset = 'eval'

eval_name = conf.get_default_eval_name()

dl = conf.get_dataloader(dset=dset, dsName=eval_name)

for batch in iter(dl):

    for k in batch.keys():
        if isinstance(batch[k], th.Tensor):
            batch[k] = batch[k].to(device)

    model_kwargs = {}

    gt = Image.open(args.get("path_y")).convert('RGB')
    data_transform=transforms.Compose([
                        transforms.ToTensor(),
                        transforms.Normalize([0.5, 0.5, 0.5], [0.5, 0.5, 0.5])
                    ])
    #print("gt.size:",gt.size)
    gt = data_transform(gt).unsqueeze(0).to("cuda")
    model_kwargs["gt"] = gt
    model_kwargs['scale'] = args.get('scale')
    model_kwargs['deg'] = args.get('deg')
    model_kwargs['resize_y'] = args.get('resize_y')
    model_kwargs['sigma_y'] = args.get('sigma_y')
    model_kwargs['save_path'] = args.get('save_path')

    gt_keep_mask = batch.get('gt_keep_mask')
    if gt_keep_mask is not None:
        model_kwargs['gt_keep_mask'] = gt_keep_mask

    batch_size = model_kwargs["gt"].shape[0]

    if conf.cond_y is not None:
        classes = th.ones(batch_size, dtype=th.long, device=device)
        model_kwargs["y"] = classes * conf.cond_y
    else:
        classes = th.randint(
            low=0, high=NUM_CLASSES, size=(batch_size,), device=device
        )
        model_kwargs["y"] = classes
        
    classes = th.ones(batch_size, dtype=th.long, device=device)
    model_kwargs["y"] = classes * args.get("class") 

    sample_fn = (
        diffusion.p_sample_loop if not conf.use_ddim else diffusion.ddim_sample_loop
    )


    result = sample_fn(
        model_fn,
        (batch_size, 3, conf.image_size, conf.image_size),
        clip_denoised=conf.clip_denoised,
        model_kwargs=model_kwargs,
        cond_fn=cond_fn,
        device=device,
        progress=show_progress,
        return_all=True,
        conf=conf
    )

print("sampling complete")
```

**Comment by wyhuai** [comment]
Thanks for your attention and greeting! 

Your generative model is of size 256x256. In this test, your input is size 128x128, and you want to get a 512x512 result. However, your generative model may be trained on aligned full brain images of size 256x256. Hence it can not handle partial brain patches of size 256x256. To solve this out-of-domain problem, you may need to train a model with random cropped brain images. Due to the same problem, models trained on standard FFHQ or CelebA-HQ datasets (they contain aligned human faces) are also improper for solving arbitrary sizes.

Besides, the 128x128 image you present seems to be aliased. It is better to use downsamplers with anti-aliasing, e.g., bicubic or average-pooling.

**Comment by JunMa11** [comment]
Hi @wyhuai ,

Thanks for your quick reply. I really appreciate your great help.

I have some follow-up questions

1. what's the correct way for downsampling?

The original image size is 256x256 and I obtain the downsampled size by `img128v1 = img[::2,::2,:]; img64v1[::4,::4,:]`.

img64v1

![ct64](https://user-images.githubusercontent.com/19947331/214599593-47f599dc-cdb6-48d3-b6ce-893f73840123.png)

img128v1

![ct128](https://user-images.githubusercontent.com/19947331/214599626-a1ec69bc-65e0-494c-b14f-30404bc11d85.png)


To address the aliased issue, I used `skimage.transform.rescale` function.
https://scikit-image.org/docs/stable/api/skimage.transform.html#skimage.transform.rescale

```python
img64v2 = transform.rescale(b_slice, 0.25, order=3, anti_aliasing=True)
img128v1 = transform.rescale(b_slice, 0.5, order=3, anti_aliasing=True)
```

img64v2 

![ct64-v2](https://user-images.githubusercontent.com/19947331/214599875-08f4bbe7-4637-4d7e-ba5b-a62bc97f01ff.png)

img128v2 

![ct128-v2](https://user-images.githubusercontent.com/19947331/214599980-4252d6d6-187f-442b-a9bf-c07c66ef28b5.png)

It seems that the aliasing issue was addressed by gaussian smoothing. In this way, I cannot verify whether the noise is removed by the model because it is already removed by gaussian smoothing. 
BTW, the noise is inherent in the image because of the imaging technique rather than hand-crafted. This is why it is important to be kept in the input image.


2. How to make the model work on `img64v1` and `img128v1`?

- For img64v1, the task is `64x64` to `256x256` and set `scale=4`
- For img64v1, the task is `128x128` to `256x256` and set `scale=2`

In this way, it doesn't have the OOD issue.


Here are the restoration results
64->256
![00000](https://user-images.githubusercontent.com/19947331/214607479-e72400a0-ea71-41d8-8254-fa5063a13ca5.png)
128->256
![00000](https://user-images.githubusercontent.com/19947331/214607564-18bf510b-458c-4f00-96b7-4d71d3c0b06e.png)

I'm also attaching the restoration results for `img64v2` and `img128v2`
64->256
![00000](https://user-images.githubusercontent.com/19947331/214607881-4b7e95fb-b1b3-4c61-ae5a-60e30602f205.png)

128->256
![00000](https://user-images.githubusercontent.com/19947331/214608005-4e6bf768-6349-425d-af1a-1d072e69ca2d.png)

The results are much better but they are not what I want because the noise is already removed in the input image.

3. How to only denoise the image? 

The input image is `256x256` and I set `scale=1; sigma_y=1.5` but it seems that the model doesn't work

Input: 

![ct256](https://user-images.githubusercontent.com/19947331/214606084-f4efa686-f493-40b9-985c-bb9e9d9264e8.png)

Output:

![00000](https://user-images.githubusercontent.com/19947331/214606176-0ec8ecb3-6c1f-443a-8f56-86222f9302ed.png)

**Comment by wyhuai** [comment]
For Q1&Q2:

Note that the operator A you used for generating the input image and the A used in DDNM should be the same. This is critical, and you may refer to Section 2.2 and Section 3.1 in our paper to understand why.

If you use `--deg "sr_averagepooling" --scale 4`,  then `A = torch.nn.AdaptiveAvgPool2d((256//scale,256//scale))` in DDNM. However, the operator in `img64v1= img[::4,::4,:]` is more like a mask operation rather than the defined A. So I suggest you use  `img64v1 = A(img)`.

For Q3:

The denoising function in hq_demo is not carefully checked, and this problem may be a bug 😂.

**Comment by vvictoryuki** [comment]
Hi @JunMa11,
Thank you for your attention to our work. I want to make some additions to @wyhuai's reply.

Let us note the original image as $\mathbf{x}\_{ori}=\mathbf{x}\_{gt}+\mathbf{n}$, where $\mathbf{x}\_{gt}$ means the clean ground truth and $\mathbf{n}$ means the imaging noise. If you want to solve a task combined with super-resolution and denoising, you create a low-resolution measurement as $\mathbf{y}=\mathbf{A}\mathbf{x}\_{ori}=\mathbf{A}(\mathbf{x}\_{gt}+\mathbf{n})=\mathbf{A}\mathbf{x}\_{gt}+\mathbf{An}$, where $\mathbf{A}$ means some down sampler.

For Q1, we do not think the down sampler denoises the original image completely, that is, $\mathbf{An}\neq \mathbf{0}$. And even if we think that the noise has mainly been removed, we do not get a high-quality image but a low-resolution image $\mathbf{A}\mathbf{x}\_{gt}$, which means that our restoration work is not really finished. 

For Q2, in our code, we produce the degraded measurements and restore them using the same degradation operator $\mathbf{A}$. However, from your statements, I guess the degradation operator $\mathbf{A}$ used in your experiments is your own defined operator, and the degradation operator used in restoration is the one provided in our code, which may cause the two operators are not the same, and DDNM does not work.

For Q3, I suggest trying different values of `sigma_y`. In experience, too small `sigma_y` will lead to noise that is not completely removed, and too large `sigma_y` will lead to blurred results. You can try to increase the value of `sigma_y` to see the effect.

**Comment by JunMa11** [comment]
Hi @wyhuai and @vvictoryuki ,

Thanks for your detailed explanations very much. The Q1 and Q2 have been well addressed. 

For Q3, I have tested different `sigma_y` but the results are the same. 

Input
![ct256](https://user-images.githubusercontent.com/19947331/215606772-b4318ccf-9249-4c09-be17-e90b0d069871.png)

sigma_y = 1.5

![00000](https://user-images.githubusercontent.com/19947331/215606864-322964d8-6e06-4824-9c84-a45b93083aa3.png)

sigma_y = 15


![00000](https://user-images.githubusercontent.com/19947331/215606978-8bcdfeb1-3a0a-4f80-b425-132989ca0a84.png)

sigma_y = 150

![00000](https://user-images.githubusercontent.com/19947331/215607010-938a270f-173f-42f7-a35e-693579e52373.png)

**Comment by JunMa11** [comment]
This issue only appears in the hq code. 

The original code does not have this issue. Thanks again for the awesome work.
